core/keyserver/epochs.go

```diff
@@ -134,48 +136,78 @@ func (*Server) ListMutationsStream(in *pb.ListMutationsRequest, stream pb.KeyTra
 	return status.Errorf(codes.Unimplemented, "ListMutationStream is unimplemented")
 }
 
-func (s *Server) logProofs(ctx context.Context, logID, firstTreeSize int64, epoch int64) (
-	*tpb.GetLatestSignedLogRootResponse,
-	*tpb.GetConsistencyProofResponse,
-	*tpb.GetInclusionProofResponse,
-	error) {
-	// Lookup log and map info.
-	logRoot, err := s.tlog.GetLatestSignedLogRoot(ctx,
-		&tpb.GetLatestSignedLogRootRequest{
-			LogId: logID,
+// logProof holds the proof for a signed map root up to signed log root.
+type logProof struct {
+	LogRoot        *tpb.SignedLogRoot
+	LogConsistency *tpb.Proof
+	LogInclusion   *tpb.Proof
+}
+
+// logProofs returns the proofs for a given epoch.
+func (s *Server) logProofs(ctx context.Context, d *domain.Domain, firstTreeSize int64, epoch int64) (*logProof, error) {
+	logRoot, logConsistency, err := s.latestLogRootProof(ctx, d, firstTreeSize)
+	if err != nil {
+		return nil, err
+	}
+
+	// Inclusion proof.
+	secondTreeSize := logRoot.GetTreeSize()
+	logInclusion, err := s.tlog.GetInclusionProof(ctx,
+		&tpb.GetInclusionProofRequest{
+			LogId: d.LogID,
+			// SignedMapRoot must be in the log at MapRevision.
+			LeafIndex: epoch,
+			TreeSize:  secondTreeSize,
 		})
 	if err != nil {
-		glog.Errorf("tlog.GetLatestSignedLogRoot(%v): %v", logID, err)
-		return nil, nil, nil, status.Error(codes.Internal, "Cannot fetch SignedLogRoot")
+		glog.Errorf("logProofs(): log.GetInclusionProof(%v, %v, %v): %v", d.LogID, epoch, secondTreeSize, err)
+		return nil, status.Error(codes.Internal, "Cannot fetch log inclusion proof")
+	}
+	return &logProof{
+		LogRoot:        logRoot,
+		LogConsistency: logConsistency,
+		LogInclusion:   logInclusion.GetProof(),
+	}, nil
+}
+
+// latestLogRootProof returns the lastest SignedLogRoot and it's consistency proof.
+func (s *Server) latestLogRootProof(ctx context.Context, d *domain.Domain, firstTreeSize int64) (*tpb.SignedLogRoot, *tpb.Proof, error) {
+
+	sth, err := s.latestLogRoot(ctx, d)
+	if err != nil {
+		return nil, nil, err
 	}
-	secondTreeSize := logRoot.GetSignedLogRoot().GetTreeSize()
 	// Consistency proof.
+	secondTreeSize := sth.GetTreeSize()
```

Comment:
Might want to short circuit if the secondTreeSize == firstTreeSize. If this can happen it should probably be tested.

Reply:
It looks like Trillian does the short circuiting logic.  Not sure if it makes sense to duplicate that logic here?  The verifying client code come from the Trillian repo, so it should be tested.

Reply:
Well you know the proof will be empty and it saves an RPC round trip. It's up to you if you think that's worth it.